File: src/chapters/numerical-calculus/content/code/trapezoid.cpp
#include <iostream>
#include <functional>
#include <cmath>
using namespace std;

// Composite trapezoidal rule for the integral of f on [a, b].
double trapezoid(function<double(double)> f, double a, double b, int n = 100) {
    double h = (b - a) / n, s = (f(a) + f(b)) / 2;
    for (int i = 1; i < n; ++i) s += f(a + i * h);
    return h * s;
}

int main() {
    cout.precision(10);
    cout << "int_0^1 e^x dx ~ " << trapezoid([](double x) { return exp(x); }, 0, 1, 100) << "\n";
}

File: src/pages/practice/guides/matlab/trapezoid.m
function I = trapezoid(f, a, b, n)
% TRAPEZOID  Composite trapezoidal rule for the integral of f on [a,b].
    if nargin < 4, n = 100; end
    x = linspace(a, b, n+1);
    y = f(x);
    h = (b - a) / n;
    I = h * (y(1)/2 + sum(y(2:end-1)) + y(end)/2);
end

% --- Demo ---
fprintf('int_0^1 e^x dx ~ %.10f\n', trapezoid(@exp, 0, 1, 100));
